// @sequelize/core/lib/dialects/sqlite/query-generator-typescript.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var query_generator_typescript_exports = {};
__export(query_generator_typescript_exports, {
  SqliteQueryGeneratorTypeScript: () => SqliteQueryGeneratorTypeScript
});
module.exports = __toCommonJS(query_generator_typescript_exports);
var import_node_crypto = require("node:crypto");
var import_check = require("../../utils/check");
var import_join_sql_fragments = require("../../utils/join-sql-fragments");
var import_string = require("../../utils/string");
var import_query_generator = require("../abstract/query-generator");
var import_query_generator_typescript = require("../abstract/query-generator-typescript");
const LIST_TABLES_QUERY_SUPPORTED_OPTIONS = /* @__PURE__ */ new Set();
const REMOVE_INDEX_QUERY_SUPPORTED_OPTIONS = /* @__PURE__ */ new Set(["ifExists"]);
class SqliteQueryGeneratorTypeScript extends import_query_generator.AbstractQueryGenerator {
  createSchemaQuery() {
    throw new Error(`Schemas are not supported in ${this.dialect.name}.`);
  }
  dropSchemaQuery() {
    throw new Error(`Schemas are not supported in ${this.dialect.name}.`);
  }
  describeTableQuery(tableName) {
    return `PRAGMA TABLE_INFO(${this.quoteTable(tableName)})`;
  }
  describeCreateTableQuery(tableName) {
    return `SELECT sql FROM sqlite_master WHERE tbl_name = ${this.escapeTable(tableName)};`;
  }
  listTablesQuery(options) {
    if (options) {
      (0, import_check.rejectInvalidOptions)(
        "listTablesQuery",
        this.dialect.name,
        import_query_generator_typescript.LIST_TABLES_QUERY_SUPPORTABLE_OPTIONS,
        LIST_TABLES_QUERY_SUPPORTED_OPTIONS,
        options
      );
    }
    return "SELECT name AS `tableName` FROM sqlite_master WHERE type='table' AND name != 'sqlite_sequence'";
  }
  showConstraintsQuery(tableName, _options) {
    return (0, import_join_sql_fragments.joinSQLFragments)([
      "SELECT sql FROM sqlite_master",
      `WHERE tbl_name = ${this.escapeTable(tableName)}`
    ]);
  }
  showIndexesQuery(tableName) {
    return `PRAGMA INDEX_LIST(${this.quoteTable(tableName)})`;
  }
  getToggleForeignKeyChecksQuery(enable) {
    return `PRAGMA foreign_keys = ${enable ? "ON" : "OFF"}`;
  }
  renameColumnQuery(_tableName, _attrNameBefore, _attrNameAfter, _attributes) {
    throw new Error(`renameColumnQuery is not supported in ${this.dialect.name}.`);
  }
  removeColumnQuery(_table, _columnName, _options) {
    throw new Error(`removeColumnQuery is not supported in ${this.dialect.name}.`);
  }
  removeIndexQuery(tableName, indexNameOrAttributes, options) {
    if (options) {
      (0, import_check.rejectInvalidOptions)(
        "removeIndexQuery",
        this.dialect.name,
        import_query_generator_typescript.REMOVE_INDEX_QUERY_SUPPORTABLE_OPTIONS,
        REMOVE_INDEX_QUERY_SUPPORTED_OPTIONS,
        options
      );
    }
    let indexName;
    if (Array.isArray(indexNameOrAttributes)) {
      const table = this.extractTableDetails(tableName);
      indexName = (0, import_string.generateIndexName)(table, { fields: indexNameOrAttributes });
    } else {
      indexName = indexNameOrAttributes;
    }
    return (0, import_join_sql_fragments.joinSQLFragments)([
      "DROP INDEX",
      options?.ifExists ? "IF EXISTS" : "",
      this.quoteIdentifier(indexName)
    ]);
  }
  // SQLite does not support renaming columns. The following is a workaround.
  _replaceColumnQuery(tableName, attrNameBefore, attrNameAfter, attributes) {
    const table = this.extractTableDetails(tableName);
    const backupTable = this.extractTableDetails(`${table.tableName}_${(0, import_node_crypto.randomBytes)(8).toString("hex")}`, table);
    const quotedTableName = this.quoteTable(table);
    const quotedBackupTableName = this.quoteTable(backupTable);
    const tableAttributes = this.attributesToSQL(attributes);
    const attributeNamesImport = Object.keys(tableAttributes).map((attr) => attrNameAfter === attr ? `${this.quoteIdentifier(attrNameBefore)} AS ${this.quoteIdentifier(attr)}` : this.quoteIdentifier(attr)).join(", ");
    const attributeNamesExport = Object.keys(tableAttributes).map((attr) => this.quoteIdentifier(attr)).join(", ");
    return [
      this.createTableQuery(backupTable, tableAttributes),
      `INSERT INTO ${quotedBackupTableName} SELECT ${attributeNamesImport} FROM ${quotedTableName};`,
      `DROP TABLE ${quotedTableName};`,
      this.createTableQuery(table, tableAttributes),
      `INSERT INTO ${quotedTableName} SELECT ${attributeNamesExport} FROM ${quotedBackupTableName};`,
      `DROP TABLE ${quotedBackupTableName};`
    ];
  }
  // SQLite has limited ALTER TABLE capapibilites which requires the below workaround involving recreating tables.
  // This leads to issues with losing data or losing foreign key references.
  _replaceTableQuery(tableName, attributes, createTableSql) {
    const table = this.extractTableDetails(tableName);
    const backupTable = this.extractTableDetails(`${table.tableName}_${(0, import_node_crypto.randomBytes)(8).toString("hex")}`, table);
    const quotedTableName = this.quoteTable(table);
    const quotedBackupTableName = this.quoteTable(backupTable);
    const tableAttributes = this.attributesToSQL(attributes);
    const attributeNames = Object.keys(tableAttributes).map((attr) => this.quoteIdentifier(attr)).join(", ");
    const backupTableSql = createTableSql ? `${createTableSql.replace(`CREATE TABLE ${quotedTableName}`, `CREATE TABLE ${quotedBackupTableName}`)};` : this.createTableQuery(backupTable, tableAttributes);
    return [
      backupTableSql,
      `INSERT INTO ${quotedBackupTableName} SELECT ${attributeNames} FROM ${quotedTableName};`,
      `DROP TABLE ${quotedTableName};`,
      `ALTER TABLE ${quotedBackupTableName} RENAME TO ${quotedTableName};`
    ];
  }
  escapeTable(tableName) {
    const table = this.extractTableDetails(tableName);
    if (table.schema) {
      return this.escape(`${table.schema}${table.delimiter}${table.tableName}`);
    }
    return this.escape(table.tableName);
  }
  versionQuery() {
    return "SELECT sqlite_version() as `version`";
  }
  tableExistsQuery(tableName) {
    return `SELECT name FROM sqlite_master WHERE type = 'table' AND name = ${this.escapeTable(tableName)}`;
  }
  /**
   * Generates an SQL query to check if there are any foreign key violations in the db schema
   *
   * @param tableName
   */
  foreignKeyCheckQuery(tableName) {
    return `PRAGMA foreign_key_check(${this.quoteTable(tableName)});`;
  }
  _addLimitAndOffset(options) {
    let fragment = "";
    if (options.limit != null) {
      fragment += ` LIMIT ${this.escape(options.limit, options)}`;
    } else if (options.offset) {
      fragment += ` LIMIT -1`;
    }
    if (options.offset) {
      fragment += ` OFFSET ${this.escape(options.offset, options)}`;
    }
    return fragment;
  }
}
